docs/reconfigurator.adoc

```diff
@@ -0,0 +1,183 @@
+* Nexus instances do not directly coordinate with each other at all.
+
+=== Execution
+
+While our approach moves a lot of tricky allocation / assignment problems to the planning step, execution brings its own complexity for two main reasons: Nexus instances can execute blueprints concurrently and any Nexus instance may be executing an old blueprint (i.e., one that _was_ the target, but is not any more).footnote:[These are unavoidable consequences of _not_ doing leader election to choose one Nexus to carry out execution.  Why not do that?  Because that creates harder problems like monitoring that Nexus, determining when it seems to have failed or become stuck, failover -- including in cases where that Nexus is _not_ stuck or failed, but merely partitioned -- etc.  This is all possible, but hard, and these code paths are not often tested in production systems.  With our approach, there is one main code path and it's frequently tested.  (Admittedly, it can still do different things depending on what's executing concurrently.)]
+
+Even when these things happen, we want that the system:
+
+* never moves backwards (i.e., towards a previous target)
+* converges towards the current target
+
+This is easier than it sounds.  Take the example of managing Omicron zones.
+
+[sidebar]
+.Example: managing Omicron zones
+--
+Reconfigurator manages the set of Omicron zones running on each sled.  How can we ensure that when changes are made, the system only moves forward even when there are multiple Nexus instances executing blueprints concurrently and some might be executing older versions?
+
+First, we apply a generation number to "the set of Omicron zones" on each sled.  Blueprints store _for each sled_ (1) the set of zones on that sled (and their configuration) and (2) the generation number.  Any time we want to change the set of zones on a sled, we make a new blueprint with the updated set of zones and the next generation number.  Execution is really simple: we make an API call to sled agent specifying the new set of zones _and_ the new generation number.  Sled Agent keeps track of the last generation number that it saw and rejects requests with an older one.  Now, if multiple Nexus instances execute the latest target, all will succeed and the first one that reaches each Sled Agent will actually update the zones on that sled.  If there's also a Nexus executing an older blueprint, it will be rejected.
+
+// TODO mermaid diagram showing concurrent execution
+--
+
+This approach can be applied to many other areas like DNS configuration, too.  Other areas (e.g., the process of updating database state to reflect internal IP allocations) sometimes require different, _ad hoc_ mechanisms.  In all cases, though, the goals are what we said above: attempting to execute a stale blueprint must never move the system backwards and as long as _something_ is executing the newer blueprint, the system should eventually get to the new target.
```

Comment:
Reading this, I'm now concerned our current executor implementation isn't correct for IP address allocation :grimacing:. I need to go back and look at this; the specific thing that came to mind is:

* Nexus A is executing an old blueprint where zone Z has an external IP address
* Nexus B is executing a new blueprint where zone Z has been expunged

I think we could get in a situation where Nexus B deallocates the IP, then Nexus A reallocates the IP, which might cause Nexus B to fail somewhere later in execution (e.g., if it wanted to assign that IP to a different zone) or might not (if that IP is not being reassigned).

This would eventually resolve itself when Nexus A caught up to the correct blueprint, but maybe the intermediate failures here are bad and this needs some rework?

Reply:
It sounds like the update to the IP allocations should be conditional on a generation number. That should fix this issue if it exists.

Reply:
I think that may be very difficult given how the current IP allocation stuff works (it's all mixed in with instance IP allocation). Maybe worth chatting live; off my head a couple ideas:

* Make the executor's IP allocation changes conditional on "the blueprint I'm executing is still the target" (within a transaction)
* Separate service external IPs from instance external IPs more fundamentally in CRDB (much more work, but maybe better long term? or maybe not if most of the external IP machinery on the networking side is the same for both?)

Reply:
I guess a hybrid third idea is to create separate, generational config for service external IPs, and a separate RPW that always tries to update the general, shared service/instance IP machinery to always reflect the latest generation?